spec/support/matchers/have_filters.rb

```diff
@@ -1,6 +1,74 @@
+# Usage in a controller spec:
+#
+# it 'includes the appropriate before_actions' do
+#   expect(subject).to have_filters(
+#     :before,
+#     :authenticate_scope!,
+#     :verify_user_is_not_second_factor_locked,
+#     :handle_two_factor_authentication,
+#     :check_already_authenticated
+#   )
+# end
+#
+# The first parameter passed to the `have_filter` method is the kind of filter,
+# such as :before for `before_action`. The rest of the parameters represent an
+# array of the expected filters.
+# If any of the filters have only: or except: options, such as:
+# before_action :check_already_authenticated, only: :new,
+# you can test it like this:
+#
+# it 'includes the appropriate before_actions' do
+#   expect(subject).to have_filters(
+#     :before,
+#     :authenticate_scope!,
+#     :verify_user_is_not_second_factor_locked,
+#     :handle_two_factor_authentication,
+#     [:check_already_authenticated, only: :new]
+#   )
+# end
+
 RSpec::Matchers.define :have_filters do |kind, *names|
   match do |controller|
-    filters = controller._process_action_callbacks.select { |f| f.kind == kind }.map(&:filter)
+    callbacks = controller._process_action_callbacks.select { |callback| callback.kind == kind }
+
+    filters = callbacks.each_with_object([]) do |f, result|
+      result << f.filter unless filter_has_only_option?(f) || filter_has_except_option?(f)
+      result << [f.filter, only: symbolized_only_action(f)] if filter_has_only_option?(f)
+      result << [f.filter, except: symbolized_except_action(f)] if filter_has_except_option?(f)
+    end
+
     names.all? { |name| filters.include?(name) }
   end
 end
+
+def filter_has_only_option?(filter)
```

Comment:
Should these be nested to avoid top level namespace pollution?  maybe it doesn't matter so much as they are in test code only?